File: 5_longest_palindromic_substring.py
class Solution:
    def expandPalindrome(self, s, left, right):
        while left > 0 and right < len(s) - 1:
            if s[left-1] != s[right + 1]:
                return left, right + 1
            left -= 1
            right += 1
        return left, right+1

    def longestPalindrome(self, s: str) -> str:
        answer = ""

        for i in range(len(s)):
            start, end = self.expandPalindrome(s, i, i)
            if len(answer) < (end - start):
                answer = s[start:end]

        for i in range(len(s)-1):
            if s[i] == s[i+1]:
                start, end = self.expandPalindrome(s, i, i+1)
                if len(answer) < (end - start):
                    answer = s[start:end]

        return answer

File: 5_longest_palindromic_substring.go
func longestPalindrome(s string) string {
    answer := ""
    for i := 0; i< len(s); i+= 1 {
        left, right := expandedPalindrome(s, i, i)
        if len(answer) < (right - left) {
            answer = s[left:right]
        }
    }

    for i := 0; i< len(s) - 1; i+= 1 {
        if s[i] == s[i+1] {
            left, right := expandedPalindrome(s, i, i+1)
            if len(answer) < (right - left) {
                answer = s[left:right]
            }
        }
    }
    return answer
}

func expandedPalindrome (str string, left int, right int) (int, int) {
    start, end := left, right
    for (start > 0) && (end < len(str) - 1) {
        if (str[start-1] != str[end+1]) {
            return start, end + 1
        }
        start -= 1
        end += 1
    }
    return start, end + 1
}
